specs/chugsplash-config.md

```diff
@@ -21,32 +21,90 @@ ChugSplash config files may be written in Javascript, Typescript, or JSON. For J
 
-TODO finish defining constructor args spec
+> Reference names are used internally by ChugSplash to identify specific instances of contracts. Each combination of a project name and contract reference name should be unique within the ChugSplash system.
 
-## Contract Variables
+# Contract Config Definition
+JSON or Javascript object defining a contract to be deployed, it's constructor args, variable definitions, and other options.
+
+Contract defintions may include the following fields:
+- contract
+- variables
+- constructorArgs
+- externalProxy
+- kind
+- previousBuildInfo
+- previousFullyQualifiedName
+- unsafe
+
+## Contract
+The name of the contract to be deployed.
+
+## Variables
+A set of state variables for the contract. See [Contract Variables](#contract-variables) for more information.
+
+## Constructor Arguments
+A set of constructro arguments for the contract. See [Contract Constructor Arguments](#contract-constructor-arguments) for more information.
+
+## External Proxy
+If this contract was initially deployed behind a proxy using a different tool. The user may define the proxy address using this field. Note that when using an external proxy, the user is expected to also define the `previousBuildInfo`, `previousFullyQualifiedName`, and `kind` fields which provide us with additional information necessary for working with external proxy types.
+
+## Kind
+An optional field defining the type of contract:
+- no-proxy: A stateless immutable contract to be deployed without a proxy
+- standard-transparent: Standard transparent proxies not deployed using OpenZeppelin.
+- oz-transparent: An OpenZeppelin transparent proxy
+- oz-ownable-uups: An OpenZeppelin UUPS proxy using Ownable
+- oz-access-control-uups: An OpenZeppelin UUPS proxy using Access Control
+
+## Previous Build Info
+A path to a build info file generated for a previous version of a contract that has already been deployed. This field needs to be defined when upgrading a contract that uses a proxy which was deployed outside of ChugSplash. Providing the previous build info allows us to run our storage slot checker
+
+## Previous Fully Qualified Name
+The fully qualified name of a contract that has been deployed outside of ChugSplash. This field needs to be defined along with the previous build info and is used to run the storage slot checker against the previous contract version.
+
+## Unsafe
+ChugSplash performs a variety of safety checks before deploying/upgrading contracts. Users may override these safety checks using this field if their situation requires it. The unsafe field is a JSON or JS object containing a set of fields which disable different safety checks including:
+- allowRenames: Allow renaming state variables
+- skipStorageCheck: Fully disable the storage slot checker
+- delegatecall: Allow use of delegatecall in implementation contracts
```

Comment:
nit: For consistency with OZ's tool, consider prepending "allow" to delegatecall, selfdestruct, and missingPublicUpgradeTo